### ISSUE TITLE
Auto resolve does not respect the Singleton definition if registered as a callback

### ISSUE THREAD

**Issue by Luc45** [lucatume/di52#15]
This test fails:

```php
/** @test */
public function it_should_respect_singletons_when_auto_resolving() {
	$c = new tad_DI52_Container;

	$c->singleton(Dependency::class, function() {
		return new Dependency;
	});

	$parent = $c->make(Depending::class);
	$parent2 = $c->make(Depending::class);

        // Failed asserting that two variables reference the same object.
	$this->assertSame($parent->getDependency(), $parent2->getDependency());

	// If not using a closure, the test would pass:
	//$c->singleton(Dependency::class, Dependency::class);
}
```

This is because `offsetGet` [calls](https://github.com/lucatume/di52/blob/master/src/tad/DI52/Container.php#L239-L241) the callback to get the instance of `Dependency`, which returns a new instance.

If the callback was executed already, it should save that instance in the singletons array and return from that.

Before I try to open a PR for it, can you please confirm this is not the expected behavior, and that test should be passing?

**Comment by lucatume** [comment]
@Luc45 you are correct, since `Dependency` is explicitly registered as a singleton it should be resolved only once, whether it's called directly by `make` or indirectly as a dependency (as in your example).

Thanks.

**Comment by lucatume** [comment]
I've changed the code a fair bit in #26, but kept the tests you put in place for this and are still passing.

**Comment by Luc45** [comment]
@lucatume  I have mentioned a test example in this issue, but I don't recall ever adding to the test suite.

I have run it on `master` and it fails. I have run it on `refactor-for-56` and it passes. I'm pushing these tests as a new PR.

**Comment by Luc45** [comment]
I believe this should be re-opened because if my tests are accurate and it indeed fails on `^2`, it's likely worth to fix it.
